analysis/Analysis/Section_3_5.lean

```diff
@@ -348,12 +348,24 @@ abbrev OrderedPair.toObject' : OrderedPair ↪ Object where
   inj' := by sorry
 
 /-- An alternate definition of a tuple, used in Exercise 3.5.2 -/
-@[ext]
 structure SetTheory.Set.Tuple (n:ℕ) where
   X: Set
   x: SetTheory.Set.Fin n → X
   surj: Function.Surjective x
 
+/-- A helper lemma so the `ext` tactic doesn't get stuck on dependent types. -/
```

Comment:
How about something like "This extensionality lemma is stated in a way that uses the correct SetTheory notions. If we put `@[ext]` on the structure, the auto-generated ext lemma is inconveniently stated using `HEq`."

That needs more editing, but I think the point is that the auto-generated ext lemma is the "wrong" one for the type. The `ext` tactic works perfectly well with it (it's not what get's stuck, is poor us who get stuck with a difficult HEq goal). It's not uncommon to need to write a better extensionality lemma.

By the way, you don't need to mention it, but just for your own knowledge `@[ext]` generates an `ext_iff` lemma from `SetTheory.Set.Tuple.ext`, so you can actually start the next exercise with `rw [SetTheory.Set.Tuple.ext_iff]`.

Reply:
That sounds good, let me play with it. I wanted to keep the note compact since it feels mostly tangential and I'm not sure if this problem comes up again.

Reply:
Btw I get the point, what I meant to convey that (from the user perspective) trying to use the ext tactic will cause them to get stuck.

Reply:
slightly reworded, thanks!